Topics/System_Design/Authentication_Service_Design.md

```diff
@@ -0,0 +1,85 @@
+
+The best option of MFA strategies is by making the user prove their *inherent characteristics*; that is, things that are unique to them that no one else can ever take from them. **Biometrics** are a good example of this; many smartphones will use fingerprints and facial recognition, as an example.
+
+### Standardized Protocols <a name="protocols"></a>
+
+Developers don't have to write everything from scratch; using a open protocol ensures that a service can have a standardized yet secure authentication mechanism. Several protocols exist (definitions from [this article](https://www.getkisi.com/blog/authentication-protocols-overview)):
+1. **[LDAP (Lightweight Directory Access Protocol)](https://www.getkisi.com/glossary/lightweight-directory-access-protocol)** allows users to "securely locate organizations, individuals, and other resources such as files and devices in a network"
+2. **[OAuth 2](https://tools.ietf.org/html/rfc6749)** is an "authorization framework that enables applications to obtain limited access to user accounts on an HTTP service, such as Facebook, GitHub, and DigitalOcean." 
+3. **[SAML (Security Assertion Markup Language)](https://saml.xml.org/)** is an "XML-based, open-standard data format for exchanging authentication and authorization data between parties, in particular, between an identity provider and a service provider." SAML is often to provide Single Sign-On (SSO) to users, which allows to access different services using only one account.
+
+## Sessions and Tokens <a name="sessions-tokens"></a>
+
+Session and token-based authentication schemes allow users to avoid having to authenticate every single time they make a request over a service.
+
+**Session-based authentication** is an older method that relies on the server to manage authentication. When a user logs in, the server generates a session, assigns a session ID stored in a browser cookie, and maintains authentication during the user's session on the site. While cookies are usually deleted upon logout, most modern browsers employ session restoring, retaining session cookies even after logout. This eliminates the need for repeated logins.
+
+**Token-based authentication** creates an *encrypted token* upon user login, granting users the ability to perform activities on the site. Unlike session IDs stored on the server, the *client* retains the token, either in memory or in a cookie.
+For session-based authentication, the server becomes complex with system or user scale. This is in contrast to token-based authentication; however, note that the latter may still involve storing API tokens in a database table for verification or revocation.
```

Comment:
Is the `This is in contrast to token-based authentication` necessary since we just covered that. The flow is a bit confusing.

Token-based authentication** creates an *encrypted token* upon user login, granting users the ability to perform activities on the site.  Unlike session IDs stored on the server, the *client* retains the token, either in memory or in a cookie. With session-based application server becomes complex with system or user scale. On the other hand, Token-based involve storing API tokens in a database table for verification or revocation.